Relax internal invariants of Dispatcher in preparation for schema-less registration

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35058 C++ smoketests for the API.
* #34668 Python-driven testing for the new operator registration API
* #34620 Minor OperatorName cleanup
* #34618 Improved error checking
* #34601 Implement schema matching inside Dispatcher
* #34456 Finish implementing schema-less impl, switch all internal registrations to it
* #34455 Add torch::jit::parseName (mirroring parseSchema/parseSchemaOrName)
* **#34221 Relax internal invariants of Dispatcher in preparation for schema-less registration**

I'm a little hesitant to put this PR up straight up for review, as we don't
get any "payoff" from this refactor, and there is still some functionality
that is missing.  However, it was getting complicated enough that this seemed
like a good place to pause for review.

The basic concept is to relax the internal invariants of Dispatcher data structures
to allow the possibility that a FunctionSchema is not specified in an Operator.
This means that you can register a bunch of implementations, and then later
register the schema (def), and have this work out.  This capability is not
exercised at all by this PR (thus, also, no new tests), but I try to build out
the capability that preserves semantics for the existing registration pattern.

In more detail:
- DispatchKeyExtractor has an uninitialized state where it doesn't look
  for dispatch keys in any arguments of the stack.
    - TODO: DispatchKeyExtractor must be able to register/deregister schema
- DispatchTable has a new constructor taking only an OperatorName for
  the uninitialized state
    - TODO: DispatchTable must be able to register/deregister schema
- OperatorDef now maintains a "weak refcount" which keeps track of the
  number of outstanding impl-only registrations (as opposed to schema
  registrations which is the strong refcount).  The strong refcount
  handles whether or not listeners get triggered, but the weak refcount
  controls whether or not we remove the operator from the map and list
  entirely.
- Replace 'registerKernel' with 'registerImpl', which is the projected
  new interface for directly registering kernels without implementations.
  The new registerImpl is written in a way that should make it possible
  to register without schema.
- Because 'registerImpl' no longer requires an OperatorHandle, change
  'registerDef' to only return a RegistrationHandleRAII.  This is marginally
  less efficient (since we're doing two hash table lookups on a registration
  now), but this won't matter in the long term, and probably doesn't
  matter now either.

Other changes:
- Rename registerBackendFallbackKernel to registerFallback (this exposed
  a bunch of places where we're improperly directly interfacing with Dispatcher;
  we need to add this capability to the true public API)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20282847](https://our.internmc.facebook.com/intern/diff/D20282847)

ezyang committed:
wip

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "wip"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Relax internal invariants of Dispatcher in preparation for schema-less registration"


I'm a little hesitant to put this PR up straight up for review, as we don't
get any "payoff" from this refactor, and there is still some functionality
that is missing.  However, it was getting complicated enough that this seemed
like a good place to pause for review.

The basic concept is to relax the internal invariants of Dispatcher data structures
to allow the possibility that a FunctionSchema is not specified in an Operator.
This means that you can register a bunch of implementations, and then later
register the schema (def), and have this work out.  This capability is not
exercised at all by this PR (thus, also, no new tests), but I try to build out
the capability that preserves semantics for the existing registration pattern.

In more detail:
- DispatchKeyExtractor has an uninitialized state where it doesn't look
  for dispatch keys in any arguments of the stack.
    - TODO: DispatchKeyExtractor must be able to register/deregister schema
- DispatchTable has a new constructor taking only an OperatorName for
  the uninitialized state
    - TODO: DispatchTable must be able to register/deregister schema
- OperatorDef now maintains a "weak refcount" which keeps track of the
  number of outstanding impl-only registrations (as opposed to schema
  registrations which is the strong refcount).  The strong refcount
  handles whether or not listeners get triggered, but the weak refcount
  controls whether or not we remove the operator from the map and list
  entirely.
- Replace 'registerKernel' with 'registerImpl', which is the projected
  new interface for directly registering kernels without implementations.
  The new registerImpl is written in a way that should make it possible
  to register without schema.
- Because 'registerImpl' no longer requires an OperatorHandle, change
  'registerDef' to only return a RegistrationHandleRAII.  This is marginally
  less efficient (since we're doing two hash table lookups on a registration
  now), but this won't matter in the long term, and probably doesn't
  matter now either.

Other changes:
- Rename registerBackendFallbackKernel to registerFallback (this exposed
  a bunch of places where we're improperly directly interfacing with Dispatcher;
  we need to add this capability to the true public API)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Relax internal invariants of Dispatcher in preparation for schema-less registration"


I'm a little hesitant to put this PR up straight up for review, as we don't
get any "payoff" from this refactor, and there is still some functionality
that is missing.  However, it was getting complicated enough that this seemed
like a good place to pause for review.

The basic concept is to relax the internal invariants of Dispatcher data structures
to allow the possibility that a FunctionSchema is not specified in an Operator.
This means that you can register a bunch of implementations, and then later
register the schema (def), and have this work out.  This capability is not
exercised at all by this PR (thus, also, no new tests), but I try to build out
the capability that preserves semantics for the existing registration pattern.

In more detail:
- DispatchKeyExtractor has an uninitialized state where it doesn't look
  for dispatch keys in any arguments of the stack.
    - TODO: DispatchKeyExtractor must be able to register/deregister schema
- DispatchTable has a new constructor taking only an OperatorName for
  the uninitialized state
    - TODO: DispatchTable must be able to register/deregister schema
- OperatorDef now maintains a "weak refcount" which keeps track of the
  number of outstanding impl-only registrations (as opposed to schema
  registrations which is the strong refcount).  The strong refcount
  handles whether or not listeners get triggered, but the weak refcount
  controls whether or not we remove the operator from the map and list
  entirely.
- Replace 'registerKernel' with 'registerImpl', which is the projected
  new interface for directly registering kernels without implementations.
  The new registerImpl is written in a way that should make it possible
  to register without schema.
- Because 'registerImpl' no longer requires an OperatorHandle, change
  'registerDef' to only return a RegistrationHandleRAII.  This is marginally
  less efficient (since we're doing two hash table lookups on a registration
  now), but this won't matter in the long term, and probably doesn't
  matter now either.

Other changes:
- Rename registerBackendFallbackKernel to registerFallback (this exposed
  a bunch of places where we're improperly directly interfacing with Dispatcher;
  we need to add this capability to the true public API)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

bhosmer reviewed (commented):
Full review needs @smessmer's eyeballs on correctness w.r.t. mem mgmt etc. but a few naming etc. suggestions inline

bhosmer reviewed (commented):
[deleted]

ezyang commented on aten/src/ATen/core/dispatch/DispatchTable.h:
```diff
@@ -89,6 +89,16 @@ class DispatchTable final {
+  : kernels_()
+  , catchallKernel_()
+  , dispatchKeyExtractor_(DispatchKeyExtractor::makeUninitialized())
+  // TODO: this toString call is a travesty
```
I don't know! It's what the code did before.

smessmer replied:
you can also store it as `OperatorName` if you want to avoid the `toString`. It's only used for error messages in here, removing the toString here would move it towards where the error message is created.

ezyang replied:
OK, I did this change in https://github.com/pytorch/pytorch/pull/34620

ezyang commented on aten/src/ATen/core/dispatch/DispatchTable.h:
```diff
@@ -165,6 +175,7 @@ class DispatchTable final {
 
   const KernelFunction* lookup(DispatchKey dispatchKey) const {
     auto& slot = kernels_[dispatchKey];
+    // TODO: this condition shouldn't be necessary
```
This is just a general TODO I noticed when writing the PR.

smessmer replied:
can you give more context? "shouldn't be necessary" as in "I expect this still works if I remove it today" or as in "we should refactor things so it isn't necessary anymore"? If the latter, give more details on what refactoring you have in mind.

ezyang replied:
We should engineer things so that the slot is always valid, and if we return the slot for something initialized, we get a "default" value that results in the correct result in dispatching.

smessmer replied:
Can you add that as a comment to the todo?

ezyang commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -86,69 +88,123 @@ OperatorHandle Dispatcher::findOrRegisterSchema_(FunctionSchema&& schema) {
+
+  auto op = findOrRegisterWithName_(op_name);
+
+  auto kernel_handle = op.operatorIterator_->op.registerKernel(dispatch_key, std::move(kernel));
```
No, it's because I'm too lazy for temporaries.

ezyang commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -38,10 +41,26 @@ class CAFFE2_API Dispatcher final {
+    // RAII destructors which will try to destruct and they had better
+    // still have a working operator handle in this case
+    size_t refcount = 0;
+    size_t weak_refcount = 0;
```
Probably :)

ezyang committed:
Update on "Relax internal invariants of Dispatcher in preparation for schema-less registration"


I'm a little hesitant to put this PR up straight up for review, as we don't
get any "payoff" from this refactor, and there is still some functionality
that is missing.  However, it was getting complicated enough that this seemed
like a good place to pause for review.

The basic concept is to relax the internal invariants of Dispatcher data structures
to allow the possibility that a FunctionSchema is not specified in an Operator.
This means that you can register a bunch of implementations, and then later
register the schema (def), and have this work out.  This capability is not
exercised at all by this PR (thus, also, no new tests), but I try to build out
the capability that preserves semantics for the existing registration pattern.

In more detail:
- DispatchKeyExtractor has an uninitialized state where it doesn't look
  for dispatch keys in any arguments of the stack.
    - TODO: DispatchKeyExtractor must be able to register/deregister schema
- DispatchTable has a new constructor taking only an OperatorName for
  the uninitialized state
    - TODO: DispatchTable must be able to register/deregister schema
- OperatorDef now maintains a "weak refcount" which keeps track of the
  number of outstanding impl-only registrations (as opposed to schema
  registrations which is the strong refcount).  The strong refcount
  handles whether or not listeners get triggered, but the weak refcount
  controls whether or not we remove the operator from the map and list
  entirely.
- Replace 'registerKernel' with 'registerImpl', which is the projected
  new interface for directly registering kernels without schemas.
  The new registerImpl is written in a way that should make it possible
  to register without schema.
- Because 'registerImpl' no longer requires an OperatorHandle, change
  'registerDef' to only return a RegistrationHandleRAII.  This is marginally
  less efficient (since we're doing two hash table lookups on a registration
  now), but this won't matter in the long term, and probably doesn't
  matter now either.

Other changes:
- Rename registerBackendFallbackKernel to registerFallback (this exposed
  a bunch of places where we're improperly directly interfacing with Dispatcher;
  we need to add this capability to the true public API)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
Renaming is easy to do, interested if there are any structural change suggestions.

ezyang committed:
Update on "Relax internal invariants of Dispatcher in preparation for schema-less registration"


I'm a little hesitant to put this PR up straight up for review, as we don't
get any "payoff" from this refactor, and there is still some functionality
that is missing.  However, it was getting complicated enough that this seemed
like a good place to pause for review.

The basic concept is to relax the internal invariants of Dispatcher data structures
to allow the possibility that a FunctionSchema is not specified in an Operator.
This means that you can register a bunch of implementations, and then later
register the schema (def), and have this work out.  This capability is not
exercised at all by this PR (thus, also, no new tests), but I try to build out
the capability that preserves semantics for the existing registration pattern.

In more detail:
- DispatchKeyExtractor has an uninitialized state where it doesn't look
  for dispatch keys in any arguments of the stack.
    - TODO: DispatchKeyExtractor must be able to register/deregister schema
- DispatchTable has a new constructor taking only an OperatorName for
  the uninitialized state
    - TODO: DispatchTable must be able to register/deregister schema
- OperatorDef now maintains a "weak refcount" which keeps track of the
  number of outstanding impl-only registrations (as opposed to schema
  registrations which is the strong refcount).  The strong refcount
  handles whether or not listeners get triggered, but the weak refcount
  controls whether or not we remove the operator from the map and list
  entirely.
- Replace 'registerKernel' with 'registerImpl', which is the projected
  new interface for directly registering kernels without schemas.
  The new registerImpl is written in a way that should make it possible
  to register without schema.
- Because 'registerImpl' no longer requires an OperatorHandle, change
  'registerDef' to only return a RegistrationHandleRAII.  This is marginally
  less efficient (since we're doing two hash table lookups on a registration
  now), but this won't matter in the long term, and probably doesn't
  matter now either.

Other changes:
- Rename registerBackendFallbackKernel to registerFallback (this exposed
  a bunch of places where we're improperly directly interfacing with Dispatcher;
  we need to add this capability to the true public API)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20282847](https://our.internmc.facebook.com/intern/diff/D20282847)

[ghstack-poisoned]

bhosmer reviewed (approved):
> Renaming is easy to do, interested if there are any structural change suggestions.

Yeah, sorry, should've been clearer - the logic change feels good to me, it simplifies things modestly and if we find that the state-folding is problematic, say w.r.t. errors, it won't be a problem to add upstream conditionals as before.

I'm mildly concerned about the risk of my missing subtle C++ disruptions due to intricacies of the original code that aren't burned into my mental model yet. Accepting to unblock, but I'd still love to get the @smessmer eye of sauron on that aspect of things

smessmer commented on aten/src/ATen/core/BackendSelectFallbackKernel.cpp:
```diff
@@ -6,7 +6,7 @@ using c10::KernelFunction;
 namespace {
 
-static auto registry = Dispatcher::singleton().registerBackendFallbackKernel(
+static auto registry = Dispatcher::singleton().registerFallback(
```
Why remove the "backend" from the name? I find it clearer to specify it, especially since we also have fallthrough and we had (or actually still have) operator level fallbacks (i.e. compound ops) too.

ezyang replied:
Two reasons:
1. Backend is not a concept that we've actually deified as a first class concept. We have plenty of "middle-end" dispatch keys which aren't actually the terminus of the chain. IIRC, "at::Backend" is still something we want to get rid of. So we shouldn't put "Backend" in the name here; at the very least, you'd have to come up with something else
2. Re confusion with other similar sounding concepts, I think the risk of confusion with operator level fallback is low, as the other two concepts integrate with the API in orthogonal ways (fallthrough is a way to implement a kernel, whereas fallback is a way you register a kernel; compound ops are the absence of a dispatch key).
3. I think Backend looks ugly and I'd prefer if all the method names are single word

bhosmer replied:
FWIW agree with Ed on (1) and (3), and would put some real weight behind the second part of (3) - especially in an API that relies (and might continue to) on chaining.

BUT to belabor (2), it's definitely super key for us to assign distinct, canonical names to these various concepts. Even just reading this thread I got confused again. 

Here are some terms we're using:
1. fallback
2. fallthrough
3. compound
4. catchAll

Personally I'd rather see us find more evocative short words and avoid e.g. prepending "backend" to (1). 

Just to recap what I think 1-4 mean, but please check:

1. a dispatch key-specific, but not op-specific, kernel the dispatcher invokes *in preference to* any op-specific kernels, when the dispatch key is matched. (A "column fill" in the dispatch table.)

2. a no-op kernel, used in conjunction with (1) and the reentrancy of the dispatcher to implement (let's call them here) hooks using dispatch keys, like BackendSelect.

3. A "row fill" in the dispatch table - per Ed above, a kernel registered for an op, but without a dispatch key. 

4. Is this the same as compound??

ezyang replied:
Yes this is correct.

bhosmer replied:
My vote would be to go with `fallback` here and commit to having a canonical glossary as part of the finished new API

ezyang replied:
I read too quickly.

> a dispatch key-specific, but not op-specific, kernel the dispatcher invokes in preference to

No, it's the opposite way. Op kernels always are taken first, only if there is no op kernel, do we go to the fallback.

smessmer commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -101,6 +101,10 @@ struct CAFFE2_API DispatchKeyExtractor final {
   }
 
+  static DispatchKeyExtractor makeUninitialized() {
+    return DispatchKeyExtractor(0);
```
This means there is no way to distinguish from an op actually taking zero arguments and an uninitialized DispatchKeyExtractor. It's probably fine though.

ezyang replied:
Yes, I don't think it causes any problems.

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -86,69 +88,123 @@ OperatorHandle Dispatcher::findOrRegisterSchema_(FunctionSchema&& schema) {
+    return *found;
+  }
+
+  operators_.emplace_back(OperatorName(op_name));
```
nit: `push_back(op_name)` is shorter than `emplace_back(OperatorName(op_name))` and has one less constructor call (the emplace_back version does call a move after you explicitly copy it while push_back only copies once and never moves)

ezyang replied:
Actually, this is not valid here, because the constructor is explicitly written to only accept rvalue references.

```
      explicit OperatorDef(FunctionSchema&& schema)
      : op(std::move(schema)) {}
    
      explicit OperatorDef(OperatorName&& op_name)
      : op(std::move(op_name)) {}
    
```

I'm not too sure why it was originally written this way; I guess they wanted us to not do a copy accidentally. A copy is needed here so I have to turn the operator name into an rvalue and then move it.

smessmer replied:
I missed that `operator_` takes `OperatorDef` and not `OperatorName`. Never mind my comment.

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -86,69 +88,123 @@ OperatorHandle Dispatcher::findOrRegisterSchema_(FunctionSchema&& schema) {
+
+  TORCH_INTERNAL_ASSERT(op.operatorIterator_->def_and_impl_count > 0);
+  --op.operatorIterator_->def_and_impl_count;
+  if (0 == op.operatorIterator_->def_and_impl_count) {
```
This should be deduplicated with the code in deregisterDef_

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -86,69 +88,123 @@ OperatorHandle Dispatcher::findOrRegisterSchema_(FunctionSchema&& schema) {
     listeners_->callOnOperatorDeregistered(op);
+  }
+  if (0 == op.operatorIterator_->def_and_impl_count) {
+    // TODO: rename this to "assert deregistration invariants"
```
sg

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -86,69 +88,123 @@ OperatorHandle Dispatcher::findOrRegisterSchema_(FunctionSchema&& schema) {
+  ++op.operatorIterator_->def_and_impl_count;
+
+  return RegistrationHandleRAII([this, op, op_name, dispatch_key, kernel_handle] {
+    op.operatorIterator_->op.deregisterKernel_(dispatch_key, kernel_handle);
```
why doesn't `deregisterKernel_` get called from inside` deregisterImpl_` instead? Would be more consistent since it's `registerImpl` which calls `registerKernel` (and you could get rid of the comment in line 171+172)

ezyang replied:
The reason is actually really dumb: `kernel_handle` is a big type (`std::list<KernelFunction>::iterator` in this PR) and if I put it in deregisterKernel then I have to add it to a bunch of extra type signatures. So I've been torn about just inlining all the deregister methods directly into the lambdas so I can reduce the amount of C++ boilerplate I have to write here.

smessmer replied:
I still feel consistency here is more important. Boilerplate can be skipped when reading it but if things behave differently than you'd expect, that can cause confusion.

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -38,10 +41,25 @@ class CAFFE2_API Dispatcher final {
+    // (in the new world, this should only ever be 1, but old style
+    // registrations may register the schema multiple times, which
+    // will increase this count).  def_and_impl_count reflects the number
+    // of combined def() and impl() registrations.  When a def() gets
```
when **the last** def() get's unregistered, we must call the deregistration listeners. Not for the ones before that.

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -125,15 +138,15 @@ class CAFFE2_API Dispatcher final {
    * fallback kernel for the given dispatch key and, if yes, call that one.
    */
-  RegistrationHandleRAII registerBackendFallbackKernel(DispatchKey dispatch_key, KernelFunction kernel);
+  RegistrationHandleRAII registerFallback(DispatchKey dispatch_key, KernelFunction kernel);
```
as mentioned above, I'd prefer having "Backend" in the name

smessmer commented on aten/src/ATen/core/dispatch/OperatorEntry.h:
```diff
@@ -33,16 +39,19 @@ class OperatorEntry final {
 
-  FunctionSchema schema_;
+  OperatorName name_;
+  c10::optional<FunctionSchema> schema_;
```
Does it have to be optional? Can we just fill in an inferred schema from the kernel function if we only have impl calls?

ezyang replied:
No, because you might have registered a boxed function and in that case there's no impl. Also it doesn't match the spec I wrote ;)

ezyang committed:
Update on "Relax internal invariants of Dispatcher in preparation for schema-less registration"


I'm a little hesitant to put this PR up straight up for review, as we don't
get any "payoff" from this refactor, and there is still some functionality
that is missing.  However, it was getting complicated enough that this seemed
like a good place to pause for review.

The basic concept is to relax the internal invariants of Dispatcher data structures
to allow the possibility that a FunctionSchema is not specified in an Operator.
This means that you can register a bunch of implementations, and then later
register the schema (def), and have this work out.  This capability is not
exercised at all by this PR (thus, also, no new tests), but I try to build out
the capability that preserves semantics for the existing registration pattern.

In more detail:
- DispatchKeyExtractor has an uninitialized state where it doesn't look
  for dispatch keys in any arguments of the stack.
    - TODO: DispatchKeyExtractor must be able to register/deregister schema
- DispatchTable has a new constructor taking only an OperatorName for
  the uninitialized state
    - TODO: DispatchTable must be able to register/deregister schema
- OperatorDef now maintains a "weak refcount" which keeps track of the
  number of outstanding impl-only registrations (as opposed to schema
  registrations which is the strong refcount).  The strong refcount
  handles whether or not listeners get triggered, but the weak refcount
  controls whether or not we remove the operator from the map and list
  entirely.
- Replace 'registerKernel' with 'registerImpl', which is the projected
  new interface for directly registering kernels without schemas.
  The new registerImpl is written in a way that should make it possible
  to register without schema.
- Because 'registerImpl' no longer requires an OperatorHandle, change
  'registerDef' to only return a RegistrationHandleRAII.  This is marginally
  less efficient (since we're doing two hash table lookups on a registration
  now), but this won't matter in the long term, and probably doesn't
  matter now either.

Other changes:
- Rename registerBackendFallbackKernel to registerFallback (this exposed
  a bunch of places where we're improperly directly interfacing with Dispatcher;
  we need to add this capability to the true public API)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20282847](https://our.internmc.facebook.com/intern/diff/D20282847)

[ghstack-poisoned]

ezyang committed:
Update on "Relax internal invariants of Dispatcher in preparation for schema-less registration"


I'm a little hesitant to put this PR up straight up for review, as we don't
get any "payoff" from this refactor, and there is still some functionality
that is missing.  However, it was getting complicated enough that this seemed
like a good place to pause for review.

The basic concept is to relax the internal invariants of Dispatcher data structures
to allow the possibility that a FunctionSchema is not specified in an Operator.
This means that you can register a bunch of implementations, and then later
register the schema (def), and have this work out.  This capability is not
exercised at all by this PR (thus, also, no new tests), but I try to build out
the capability that preserves semantics for the existing registration pattern.

In more detail:
- DispatchKeyExtractor has an uninitialized state where it doesn't look
  for dispatch keys in any arguments of the stack.
    - TODO: DispatchKeyExtractor must be able to register/deregister schema
- DispatchTable has a new constructor taking only an OperatorName for
  the uninitialized state
    - TODO: DispatchTable must be able to register/deregister schema
- OperatorDef now maintains a "weak refcount" which keeps track of the
  number of outstanding impl-only registrations (as opposed to schema
  registrations which is the strong refcount).  The strong refcount
  handles whether or not listeners get triggered, but the weak refcount
  controls whether or not we remove the operator from the map and list
  entirely.
- Replace 'registerKernel' with 'registerImpl', which is the projected
  new interface for directly registering kernels without implementations.
  The new registerImpl is written in a way that should make it possible
  to register without schema.
- Because 'registerImpl' no longer requires an OperatorHandle, change
  'registerDef' to only return a RegistrationHandleRAII.  This is marginally
  less efficient (since we're doing two hash table lookups on a registration
  now), but this won't matter in the long term, and probably doesn't
  matter now either.

Other changes:
- Rename registerBackendFallbackKernel to registerFallback (this exposed
  a bunch of places where we're improperly directly interfacing with Dispatcher;
  we need to add this capability to the true public API)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20282847](https://our.internmc.facebook.com/intern/diff/D20282847)

[ghstack-poisoned]

ezyang committed:
Update on "Relax internal invariants of Dispatcher in preparation for schema-less registration"


I'm a little hesitant to put this PR up straight up for review, as we don't
get any "payoff" from this refactor, and there is still some functionality
that is missing.  However, it was getting complicated enough that this seemed
like a good place to pause for review.

The basic concept is to relax the internal invariants of Dispatcher data structures
to allow the possibility that a FunctionSchema is not specified in an Operator.
This means that you can register a bunch of implementations, and then later
register the schema (def), and have this work out.  This capability is not
exercised at all by this PR (thus, also, no new tests), but I try to build out
the capability that preserves semantics for the existing registration pattern.

In more detail:
- DispatchKeyExtractor has an uninitialized state where it doesn't look
  for dispatch keys in any arguments of the stack.
    - TODO: DispatchKeyExtractor must be able to register/deregister schema
- DispatchTable has a new constructor taking only an OperatorName for
  the uninitialized state
    - TODO: DispatchTable must be able to register/deregister schema
- OperatorDef now maintains a "weak refcount" which keeps track of the
  number of outstanding impl-only registrations (as opposed to schema
  registrations which is the strong refcount).  The strong refcount
  handles whether or not listeners get triggered, but the weak refcount
  controls whether or not we remove the operator from the map and list
  entirely.
- Replace 'registerKernel' with 'registerImpl', which is the projected
  new interface for directly registering kernels without implementations.
  The new registerImpl is written in a way that should make it possible
  to register without schema.
- Because 'registerImpl' no longer requires an OperatorHandle, change
  'registerDef' to only return a RegistrationHandleRAII.  This is marginally
  less efficient (since we're doing two hash table lookups on a registration
  now), but this won't matter in the long term, and probably doesn't
  matter now either.

Other changes:
- Rename registerBackendFallbackKernel to registerFallback (this exposed
  a bunch of places where we're improperly directly interfacing with Dispatcher;
  we need to add this capability to the true public API)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20282847](https://our.internmc.facebook.com/intern/diff/D20282847)

[ghstack-poisoned]

ezyang committed:
Update on "Relax internal invariants of Dispatcher in preparation for schema-less registration"


I'm a little hesitant to put this PR up straight up for review, as we don't
get any "payoff" from this refactor, and there is still some functionality
that is missing.  However, it was getting complicated enough that this seemed
like a good place to pause for review.

The basic concept is to relax the internal invariants of Dispatcher data structures
to allow the possibility that a FunctionSchema is not specified in an Operator.
This means that you can register a bunch of implementations, and then later
register the schema (def), and have this work out.  This capability is not
exercised at all by this PR (thus, also, no new tests), but I try to build out
the capability that preserves semantics for the existing registration pattern.

In more detail:
- DispatchKeyExtractor has an uninitialized state where it doesn't look
  for dispatch keys in any arguments of the stack.
    - TODO: DispatchKeyExtractor must be able to register/deregister schema
- DispatchTable has a new constructor taking only an OperatorName for
  the uninitialized state
    - TODO: DispatchTable must be able to register/deregister schema
- OperatorDef now maintains a "weak refcount" which keeps track of the
  number of outstanding impl-only registrations (as opposed to schema
  registrations which is the strong refcount).  The strong refcount
  handles whether or not listeners get triggered, but the weak refcount
  controls whether or not we remove the operator from the map and list
  entirely.
- Replace 'registerKernel' with 'registerImpl', which is the projected
  new interface for directly registering kernels without implementations.
  The new registerImpl is written in a way that should make it possible
  to register without schema.
- Because 'registerImpl' no longer requires an OperatorHandle, change
  'registerDef' to only return a RegistrationHandleRAII.  This is marginally
  less efficient (since we're doing two hash table lookups on a registration
  now), but this won't matter in the long term, and probably doesn't
  matter now either.

Other changes:
- Rename registerBackendFallbackKernel to registerFallback (this exposed
  a bunch of places where we're improperly directly interfacing with Dispatcher;
  we need to add this capability to the true public API)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20282847](https://our.internmc.facebook.com/intern/diff/D20282847)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -57,7 +57,9 @@ OperatorHandle Dispatcher::findSchemaOrThrow(const char* name, const char* overl
-OperatorHandle Dispatcher::findOrRegisterSchema_(FunctionSchema&& schema) {
+// Postcondition: caller is responsible for disposing of registration when they
+// are done
+OperatorHandle Dispatcher::findOrRegisterWithSchema_(FunctionSchema&& schema) {
```
This function got eliminated entirely as the refactoring proceed further

ezyang committed:
Update on "Relax internal invariants of Dispatcher in preparation for schema-less registration"


I'm a little hesitant to put this PR up straight up for review, as we don't
get any "payoff" from this refactor, and there is still some functionality
that is missing.  However, it was getting complicated enough that this seemed
like a good place to pause for review.

The basic concept is to relax the internal invariants of Dispatcher data structures
to allow the possibility that a FunctionSchema is not specified in an Operator.
This means that you can register a bunch of implementations, and then later
register the schema (def), and have this work out.  This capability is not
exercised at all by this PR (thus, also, no new tests), but I try to build out
the capability that preserves semantics for the existing registration pattern.

In more detail:
- DispatchKeyExtractor has an uninitialized state where it doesn't look
  for dispatch keys in any arguments of the stack.
    - TODO: DispatchKeyExtractor must be able to register/deregister schema
- DispatchTable has a new constructor taking only an OperatorName for
  the uninitialized state
    - TODO: DispatchTable must be able to register/deregister schema
- OperatorDef now maintains a "weak refcount" which keeps track of the
  number of outstanding impl-only registrations (as opposed to schema
  registrations which is the strong refcount).  The strong refcount
  handles whether or not listeners get triggered, but the weak refcount
  controls whether or not we remove the operator from the map and list
  entirely.
- Replace 'registerKernel' with 'registerImpl', which is the projected
  new interface for directly registering kernels without implementations.
  The new registerImpl is written in a way that should make it possible
  to register without schema.
- Because 'registerImpl' no longer requires an OperatorHandle, change
  'registerDef' to only return a RegistrationHandleRAII.  This is marginally
  less efficient (since we're doing two hash table lookups on a registration
  now), but this won't matter in the long term, and probably doesn't
  matter now either.

Other changes:
- Rename registerBackendFallbackKernel to registerFallback (this exposed
  a bunch of places where we're improperly directly interfacing with Dispatcher;
  we need to add this capability to the true public API)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20282847](https://our.internmc.facebook.com/intern/diff/D20282847)

[ghstack-poisoned]

ezyang committed:
Update on "Relax internal invariants of Dispatcher in preparation for schema-less registration"


I'm a little hesitant to put this PR up straight up for review, as we don't
get any "payoff" from this refactor, and there is still some functionality
that is missing.  However, it was getting complicated enough that this seemed
like a good place to pause for review.

The basic concept is to relax the internal invariants of Dispatcher data structures
to allow the possibility that a FunctionSchema is not specified in an Operator.
This means that you can register a bunch of implementations, and then later
register the schema (def), and have this work out.  This capability is not
exercised at all by this PR (thus, also, no new tests), but I try to build out
the capability that preserves semantics for the existing registration pattern.

In more detail:
- DispatchKeyExtractor has an uninitialized state where it doesn't look
  for dispatch keys in any arguments of the stack.
    - TODO: DispatchKeyExtractor must be able to register/deregister schema
- DispatchTable has a new constructor taking only an OperatorName for
  the uninitialized state
    - TODO: DispatchTable must be able to register/deregister schema
- OperatorDef now maintains a "weak refcount" which keeps track of the
  number of outstanding impl-only registrations (as opposed to schema
  registrations which is the strong refcount).  The strong refcount
  handles whether or not listeners get triggered, but the weak refcount
  controls whether or not we remove the operator from the map and list
  entirely.
- Replace 'registerKernel' with 'registerImpl', which is the projected
  new interface for directly registering kernels without implementations.
  The new registerImpl is written in a way that should make it possible
  to register without schema.
- Because 'registerImpl' no longer requires an OperatorHandle, change
  'registerDef' to only return a RegistrationHandleRAII.  This is marginally
  less efficient (since we're doing two hash table lookups on a registration
  now), but this won't matter in the long term, and probably doesn't
  matter now either.

Other changes:
- Rename registerBackendFallbackKernel to registerFallback (this exposed
  a bunch of places where we're improperly directly interfacing with Dispatcher;
  we need to add this capability to the true public API)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20282847](https://our.internmc.facebook.com/intern/diff/D20282847)

[ghstack-poisoned]

ezyang committed:
Update on "Relax internal invariants of Dispatcher in preparation for schema-less registration"


I'm a little hesitant to put this PR up straight up for review, as we don't
get any "payoff" from this refactor, and there is still some functionality
that is missing.  However, it was getting complicated enough that this seemed
like a good place to pause for review.

The basic concept is to relax the internal invariants of Dispatcher data structures
to allow the possibility that a FunctionSchema is not specified in an Operator.
This means that you can register a bunch of implementations, and then later
register the schema (def), and have this work out.  This capability is not
exercised at all by this PR (thus, also, no new tests), but I try to build out
the capability that preserves semantics for the existing registration pattern.

In more detail:
- DispatchKeyExtractor has an uninitialized state where it doesn't look
  for dispatch keys in any arguments of the stack.
    - TODO: DispatchKeyExtractor must be able to register/deregister schema
- DispatchTable has a new constructor taking only an OperatorName for
  the uninitialized state
    - TODO: DispatchTable must be able to register/deregister schema
- OperatorDef now maintains a "weak refcount" which keeps track of the
  number of outstanding impl-only registrations (as opposed to schema
  registrations which is the strong refcount).  The strong refcount
  handles whether or not listeners get triggered, but the weak refcount
  controls whether or not we remove the operator from the map and list
  entirely.
- Replace 'registerKernel' with 'registerImpl', which is the projected
  new interface for directly registering kernels without implementations.
  The new registerImpl is written in a way that should make it possible
  to register without schema.
- Because 'registerImpl' no longer requires an OperatorHandle, change
  'registerDef' to only return a RegistrationHandleRAII.  This is marginally
  less efficient (since we're doing two hash table lookups on a registration
  now), but this won't matter in the long term, and probably doesn't
  matter now either.

Other changes:
- Rename registerBackendFallbackKernel to registerFallback (this exposed
  a bunch of places where we're improperly directly interfacing with Dispatcher;
  we need to add this capability to the true public API)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20282847](https://our.internmc.facebook.com/intern/diff/D20282847)

[ghstack-poisoned]

ezyang committed:
Update on "Relax internal invariants of Dispatcher in preparation for schema-less registration"


I'm a little hesitant to put this PR up straight up for review, as we don't
get any "payoff" from this refactor, and there is still some functionality
that is missing.  However, it was getting complicated enough that this seemed
like a good place to pause for review.

The basic concept is to relax the internal invariants of Dispatcher data structures
to allow the possibility that a FunctionSchema is not specified in an Operator.
This means that you can register a bunch of implementations, and then later
register the schema (def), and have this work out.  This capability is not
exercised at all by this PR (thus, also, no new tests), but I try to build out
the capability that preserves semantics for the existing registration pattern.

In more detail:
- DispatchKeyExtractor has an uninitialized state where it doesn't look
  for dispatch keys in any arguments of the stack.
    - TODO: DispatchKeyExtractor must be able to register/deregister schema
- DispatchTable has a new constructor taking only an OperatorName for
  the uninitialized state
    - TODO: DispatchTable must be able to register/deregister schema
- OperatorDef now maintains a "weak refcount" which keeps track of the
  number of outstanding impl-only registrations (as opposed to schema
  registrations which is the strong refcount).  The strong refcount
  handles whether or not listeners get triggered, but the weak refcount
  controls whether or not we remove the operator from the map and list
  entirely.
- Replace 'registerKernel' with 'registerImpl', which is the projected
  new interface for directly registering kernels without implementations.
  The new registerImpl is written in a way that should make it possible
  to register without schema.
- Because 'registerImpl' no longer requires an OperatorHandle, change
  'registerDef' to only return a RegistrationHandleRAII.  This is marginally
  less efficient (since we're doing two hash table lookups on a registration
  now), but this won't matter in the long term, and probably doesn't
  matter now either.

Other changes:
- Rename registerBackendFallbackKernel to registerFallback (this exposed
  a bunch of places where we're improperly directly interfacing with Dispatcher;
  we need to add this capability to the true public API)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20282847](https://our.internmc.facebook.com/intern/diff/D20282847)

[ghstack-poisoned]

ezyang committed:
Update on "Relax internal invariants of Dispatcher in preparation for schema-less registration"


I'm a little hesitant to put this PR up straight up for review, as we don't
get any "payoff" from this refactor, and there is still some functionality
that is missing.  However, it was getting complicated enough that this seemed
like a good place to pause for review.

The basic concept is to relax the internal invariants of Dispatcher data structures
to allow the possibility that a FunctionSchema is not specified in an Operator.
This means that you can register a bunch of implementations, and then later
register the schema (def), and have this work out.  This capability is not
exercised at all by this PR (thus, also, no new tests), but I try to build out
the capability that preserves semantics for the existing registration pattern.

In more detail:
- DispatchKeyExtractor has an uninitialized state where it doesn't look
  for dispatch keys in any arguments of the stack.
    - TODO: DispatchKeyExtractor must be able to register/deregister schema
- DispatchTable has a new constructor taking only an OperatorName for
  the uninitialized state
    - TODO: DispatchTable must be able to register/deregister schema
- OperatorDef now maintains a "weak refcount" which keeps track of the
  number of outstanding impl-only registrations (as opposed to schema
  registrations which is the strong refcount).  The strong refcount
  handles whether or not listeners get triggered, but the weak refcount
  controls whether or not we remove the operator from the map and list
  entirely.
- Replace 'registerKernel' with 'registerImpl', which is the projected
  new interface for directly registering kernels without implementations.
  The new registerImpl is written in a way that should make it possible
  to register without schema.
- Because 'registerImpl' no longer requires an OperatorHandle, change
  'registerDef' to only return a RegistrationHandleRAII.  This is marginally
  less efficient (since we're doing two hash table lookups on a registration
  now), but this won't matter in the long term, and probably doesn't
  matter now either.

Other changes:
- Rename registerBackendFallbackKernel to registerFallback (this exposed
  a bunch of places where we're improperly directly interfacing with Dispatcher;
  we need to add this capability to the true public API)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20282847](https://our.internmc.facebook.com/intern/diff/D20282847)

[ghstack-poisoned]

ezyang commented:
All the CR comments have been fully addressed and pushed.

ezyang committed:
Update on "Relax internal invariants of Dispatcher in preparation for schema-less registration"


I'm a little hesitant to put this PR up straight up for review, as we don't
get any "payoff" from this refactor, and there is still some functionality
that is missing.  However, it was getting complicated enough that this seemed
like a good place to pause for review.

The basic concept is to relax the internal invariants of Dispatcher data structures
to allow the possibility that a FunctionSchema is not specified in an Operator.
This means that you can register a bunch of implementations, and then later
register the schema (def), and have this work out.  This capability is not
exercised at all by this PR (thus, also, no new tests), but I try to build out
the capability that preserves semantics for the existing registration pattern.

In more detail:
- DispatchKeyExtractor has an uninitialized state where it doesn't look
  for dispatch keys in any arguments of the stack.
    - TODO: DispatchKeyExtractor must be able to register/deregister schema
- DispatchTable has a new constructor taking only an OperatorName for
  the uninitialized state
    - TODO: DispatchTable must be able to register/deregister schema
- OperatorDef now maintains a "weak refcount" which keeps track of the
  number of outstanding impl-only registrations (as opposed to schema
  registrations which is the strong refcount).  The strong refcount
  handles whether or not listeners get triggered, but the weak refcount
  controls whether or not we remove the operator from the map and list
  entirely.
- Replace 'registerKernel' with 'registerImpl', which is the projected
  new interface for directly registering kernels without implementations.
  The new registerImpl is written in a way that should make it possible
  to register without schema.
- Because 'registerImpl' no longer requires an OperatorHandle, change
  'registerDef' to only return a RegistrationHandleRAII.  This is marginally
  less efficient (since we're doing two hash table lookups on a registration
  now), but this won't matter in the long term, and probably doesn't
  matter now either.

Other changes:
- Rename registerBackendFallbackKernel to registerFallback (this exposed
  a bunch of places where we're improperly directly interfacing with Dispatcher;
  we need to add this capability to the true public API)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20282847](https://our.internmc.facebook.com/intern/diff/D20282847)

[ghstack-poisoned]
